Use correlated searching for repeated interpolations

moble opened the pull request:
As suggested in #142, this PR adds the ability to track the last index found when interpolating, and use it as a guess for the next interpolation.  Essentially, the current use of `searchsortedlast`/`first` starts over, doing a binary seach over the entire `t` array every time the interpolator is called.  If we track the last index found, then the next time the interpolator is called, we can first do an *expanding* binary search to bracket the desired value, then call `searchsortedlast`/`first` as usual.  When successive values are close together, this means that the searches are much faster.

While the cost of constructing the interpolator does not change, tracking the last index results in a best-case speedup of ~2.7x for `CubicSpline`, when successive values are close together (and a little higher for simpler interpolators).  In the worst case (where successive values are always on opposite ends of the vectors), it can result in a ~15% slowdown due to the unhelpful expanding binary search at the beginning.  However, the original approach of not tracking the index at all is also still available; it now involves essentially one extra `if` statement, which seems to be lost in the timing noise.

The interface — both the public and `_interpolate` parts — should be backwards compatible, but have optional extra arguments for the index.  I've also added a method `(interp::AbstractInterpolation)(t::AbstractVector)`, along with an in-place mutating version, which are features that I have felt were sorely lacking.

I extended this tracking to the `derivative` interface, but left the `integral` interface alone because I don't know that it would be much of an improvement, and it starts to get a little complicated.  There were also a few things like `LagrangeInterpolation` that I left alone because they use a different approach.

ChrisRackauckas commented on src/interpolation_utils.jl:
```diff
@@ -57,3 +57,68 @@ function munge_data(U::StridedMatrix, t::AbstractVector)
+  return lo, hi
+end
+
+for s in [:searchsortedfirst, :searchsortedlast]
+  @eval begin
+    function Base.$s(v::AbstractVector, x, guess)
```
Extending `searchsortedfirst` is type piracy. We should just make a new function.

moble replied:
Renamed to `searchsortedfirstcorrelated` (and `...last...`)

ChrisRackauckas commented:
This looks good to go, just fix the merge conflict (it moved files) and this can merge. It's breaking because it changes the derivative handling, but this will get merged into the v4 release which is building up for next week, so that's totally fine.

moble commented:
Thanks.  I've rebased and added some description to the README.

For reference, [here's the script](https://gist.github.com/moble/8c3c3369324f0597045db09a6575ad35) I used to benchmark the new approach versus the old.

moble commented:
Also, I think this can close #52 and #142.
